File: src/main/java/com/github/pnemonic/game/lottery/pais777/Lotto777.java
package com.github.pnemonic.game.lottery.pais777;

import com.github.pnemonic.game.lottery.LotException;
import com.github.pnemonic.game.lottery.Lottery;
import com.github.pnemonic.game.lottery.LotteryGame;

import java.util.List;
import java.util.Set;

/**
 * Choose numbers for 777 for Israel.
 * 
 * @author Moshe
 */
public class Lotto777 extends Lottery {

	/** Number of groupings to play. */
	private static final int GROUPINGS = 1;
	/** Cost per game. */
	public static final double COST = 7.00;
	/** Budget. */
	public static final double BUDGET = 200.00 / GROUPINGS;
	/** Total number of plays per budget. */
	public static final int PLAYS = (int) Math.floor(BUDGET / COST);

	private static final int SIZE = 7;
	private static final int MAX_LOWER = 23;
	private static final int MIN_UPPER = 43;
	private static final int MAX_GAP = 32;
	private static final int MIN_ODD = 2;
	private static final int MIN_EVEN = 2;

	/**
	 * Constructs a new 777.
	 */
	public Lotto777() {
		super();
	}

	/**
	 * Main method.
	 * 
	 * @param args
	 *            the array of arguments.
	 */
	public static void main(String[] args) {
		Lottery lottery = new Lotto777();
		if (args.length > 0) {
			lottery.setCandidates(args[0]);
		}

		int plays = PLAYS;
		Set<LotteryGame> games = lottery.play(plays);
		for (LotteryGame game : games) {
			lottery.print(game);
		}
	}

	@Override
	public int size() {
		return SIZE;
	}

	@Override
	public int getMinimum() {
		return 1;
	}

	@Override
	public int getMaximum() {
		return 70;
	}

	@Override
	public int getBonusMinimum() {
		return 0;
	}

	@Override
	public int getBonusMaximum() {
		return 0;
	}

	@Override
	protected void filter(LotteryGame game, int pickIndex, List<Integer> bag) throws LotException {
		super.filter(game, pickIndex, bag);

		// Rule: Too few "5 consecutive" numbers.
		if (pickIndex >= 4) {
			int[] lot = game.lot;
			int candidate;

			if (pickIndex >= 4) {
				int i = pickIndex;
				if ((lot[i - 4] + 1 == lot[i - 3]) && (lot[i - 3] + 1 == lot[i - 2]) && (lot[i - 2] + 1 == lot[i - 1]) && (lot[i - 1] + 1 == lot[i])) {
					// remove 5th consecutive numbers: 1 before; 1 after.
					candidate = lot[i - 4] - 1;
					bag.remove((Object) candidate);
					candidate = lot[i] + 1;
					bag.remove((Object) candidate);
				}
			}
		}
	}

	@Override
	protected void filterGame(LotteryGame game) throws LotException {
		super.filterGame(game);
		int candidate;
		int[] lot = game.lot;
		int size = lot.length;

		// Rule: Minimum-valued ball cannot exceed MAX_LOWER.
		candidate = lot[0];
		if (candidate > MAX_LOWER) {
			throw new LotException("Minimum-valued ball " + candidate + " exceeds " + MAX_LOWER);
		}
		// Rule: Maximum-valued ball cannot precede MIN_UPPER.
		candidate = lot[size - 1];
		if (candidate < MIN_UPPER) {
			throw new LotException("Maximum-valued ball " + candidate + " less than " + MIN_UPPER);
		}

		// Rule: Biggest gap between 2 balls.
		for (int l = 0, l1 = 1; l1 < size; l++, l1++) {
			if (lot[l] + MAX_GAP <= lot[l1]) {
				throw new LotException("Widest gap between " + lot[l] + " and " + lot[l1] + " exceeds " + MAX_GAP);
			}
		}

		// Rule: Minimum number of odd-numbered balls.
		int countOdd = 0;
		int countEven = 0;
		for (int l : lot) {
			if ((l & 1) == 0) {
				countEven++;
			} else {
				countOdd++;
			}
		}
		if (countOdd < MIN_ODD) {
			throw new LotException("Minimum odds less than " + MIN_ODD);
		}
		if (countEven < MIN_EVEN) {
			throw new LotException("Minimum evens less than " + MIN_EVEN);
		}
	}

}

File: src/main/java/com/github/pnemonic/game/lottery/pais777/Lotto777.kt
package com.github.pnemonic.game.lottery.pais777

import com.github.pnemonic.game.GameException
import com.github.pnemonic.game.lottery.Lottery
import com.github.pnemonic.game.lottery.LotteryGame
import com.github.pnemonic.game.lottery.LotteryGuess
import com.github.pnemonic.game.lottery.pais777.Lotto777.Companion.COST
import com.github.pnemonic.game.lottery.pais777.Lotto777Tester.Companion.BUDGET
import com.github.pnemonic.isEven

/**
 * Choose numbers for 777 in Israel.
 */
open class Lotto777 : Lottery(SIZE) {
    override val minimum: Int = 1
    override val maximum: Int = 70
    override val bonusMinimum: Int = 0
    override val bonusMaximum: Int = 0

    @Throws(GameException::class)
    override fun filterGame(game: LotteryGame) {
        super.filterGame(game)
        var candidate: Int
        val lot = game.balls
        val size = lot.size

        // Rule: Minimum-valued ball cannot exceed MAX_LOWER.
        candidate = lot[0]
        if (candidate > MAX_LOWER) {
            throw GameException("Minimum-valued ball $candidate exceeds $MAX_LOWER")
        }
        // Rule: Maximum-valued ball cannot precede MIN_UPPER.
        candidate = lot[size - 1]
        if (candidate < MIN_UPPER) {
            throw GameException("Maximum-valued ball $candidate less than $MIN_UPPER")
        }

        // Rule: Minimum number of odd-numbered balls.
        var countOdd = 0
        var countEven = 0
        for (l in lot) {
            if (l.isEven) {
                countEven++
            } else {
                countOdd++
            }
        }
        if (countOdd < MIN_ODD) {
            throw GameException("Odds less than $MIN_ODD")
        }
        if (countEven < MIN_EVEN) {
            throw GameException("Evens less than $MIN_EVEN")
        }
    }

    override fun calculatePrizes(guess: LotteryGuess, result: LotteryGame) {
        var match = 0
        for (ball in guess.balls) {
            if (result.balls.binarySearch(ball) >= 0) {
                match++
            }
        }

        when (match) {
            3 -> result.prize = PRIZE_3
            4 -> result.prize = PRIZE_4
            5 -> result.prize = PRIZE_5
            6 -> result.prize = PRIZE_6
            7 -> result.prize = PRIZE_7
            else -> result.prize = 0
        }
    }

    companion object {
        private const val SIZE = 7
        private const val MAX_LOWER = 23
        private const val MIN_UPPER = 43
        private const val MIN_ODD = 2
        private const val MIN_EVEN = 2

        /**
         * Cost per game.
         */
        const val COST = 7

        /** Prize for guessing 3 correct numbers. */
        const val PRIZE_3 = 5

        /** Prize for guessing 4 correct numbers. */
        const val PRIZE_4 = 20

        /** Prize for guessing 5 correct numbers. */
        const val PRIZE_5 = 50

        /** Prize for guessing 6 correct numbers. */
        const val PRIZE_6 = 500

        /** Prize for guessing all 7 correct numbers. */
        const val PRIZE_7 = 70_000
    }
}

/**
 * Main method.
 *
 * @param args the array of arguments.
 */
fun main(args: Array<String>) {
    //TODO parse history then predict some guesses
    val lottery: Lottery = Lotto777()
    if (args.isNotEmpty()) {
        lottery.setCandidates(args[0])
    }
    val numPlays = BUDGET / COST
    val games = lottery.play(numPlays)
    for (game in games) {
        lottery.print(game)
    }
}